### PR TITLE
KYLIN-4813 Refactor spark build log

### PR DESCRIPTION
**Pull request by zhangayqian**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [ ] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I will prepare another pr against the `document` branch
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin or dev@kylin by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 1]
```diff
@@ -2803,10 +2822,15 @@ public boolean isSparkEngineEnabled() {
...
-        return getLogPropertyFile("kylin-parquet-log4j.properties");
+    public String getLogSparkDriverPropertiesFile() {
+        return getLogPropertyFile("spark-driver-log4j.properties");
+    }
+
+    public String getLogSparkExecutorPropertiesFile() {
+        return getLogPropertyFile("spark-executor-log4j.properties");
     }
 
+
```
Comment (zzcclp) [inline core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:2833]: remove this blank line.

File: server-base/src/main/java/org/apache/kylin/rest/controller/JobController.java [hunk 2]
```diff
@@ -160,17 +168,41 @@ public JobInstance get(@PathVariable String jobId) {
...
+     */
+    @RequestMapping(value = "/{job_id:.+}/steps/{step_id:.+}/log", method = { RequestMethod.GET }, produces = { "application/json" })
+    @ResponseBody
+    public EnvelopeResponse<String> downloadLogFile(@PathVariable("job_id") String jobId,
+                                                    @PathVariable("step_id") String stepId, @RequestParam(value = "project") String project,
+                                                    HttpServletResponse response) throws IOException {
+        checkRequiredArg("job_id", jobId);
+        checkRequiredArg("step_id", stepId);
+        checkRequiredArg("project", project);
+        String downloadFilename = String.format(Locale.ROOT, "%s_%s.log", project, stepId);
```
Comment (hit-lacus) [inline server-base/src/main/java/org/apache/kylin/rest/controller/JobController.java:199]: I am afraid this will introduce security issue.

Comment (hit-lacus) [inline server-base/src/main/java/org/apache/kylin/rest/controller/JobController.java:199]: Before combine `project` with `stepId`, please check if it is vaild and remove any illegal character.

Comment (hit-lacus) [reply server-base/src/main/java/org/apache/kylin/rest/controller/JobController.java:199]: Please check report from LGTM or search "HTTP Response Splitting" for background.

File: core-common/src/main/resources/kylin-defaults.properties [hunk 3]
```diff
@@ -283,6 +287,7 @@ kylin.query.spark-conf.spark.serializer=org.apache.spark.serializer.JavaSerializ
 #kylin.query.spark-conf.spark.sql.shuffle.partitions=40
 #kylin.query.spark-conf.spark.yarn.jars=hdfs://localhost:9000/spark2_jars/*
 
+kylin.storage.columnar.spark-conf.spark.executor.extraJavaOptions=-Dhdp.version=current -Dlog4j.configuration=spark-executor-log4j.properties -Dlog4j.debug -Dkylin.hdfs.working.dir=${kylin.env.hdfs-working-dir} -Dkylin.metadata.identifier=${kylin.metadata.url.identifier} -Dkylin.spark.category=sparder -Dkylin.spark.project=${job.project} -XX:MaxDirectMemorySize=896M
```
Comment (hit-lacus) [inline core-common/src/main/resources/kylin-defaults.properties:290]: `kylin.engine.spark-conf` or `kylin.engine.spark-conf` ?

File: core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 4]
```diff
@@ -258,6 +258,9 @@ final public void setProperty(String key, String value) {
 
     final protected void reloadKylinConfig(Properties properties) {
         this.properties = BCC.check(properties);
+        setProperty("kylin.metadata.url.identifier", getMetadataUrlPrefix());
+        setProperty("kylin.log.spark-driver-properties-file", getLogSparkDriverPropertiesFile());
```
Comment (hit-lacus) [inline core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java:262]: What does `kylin.log.spark-driver-properties-file` means?

File: core-common/src/main/resources/kylin-defaults.properties (hunk 3 above)
Comment (zzcclp) [reply core-common/src/main/resources/kylin-defaults.properties:290]: Please replace 'kylin.storage.columnar' with 'kylin.query'.

File: build/conf/spark-executor-log4j.properties [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+log4j.rootLogger=INFO,stderr,hdfs
+
+log4j.appender.stderr=org.apache.log4j.ConsoleAppender
+log4j.appender.stderr.layout=org.apache.kylin.common.logging.SensitivePatternLayout
+log4j.appender.stderr.target=System.err
+#Don't add line number (%L) as it's too costly!
+log4j.appender.stderr.layout.ConversionPattern=%d{ISO8601} %-5p [%t] %c{2} : %m%n
+
+
+log4j.appender.hdfs=org.apache.kylin.engine.spark.common.logging.SparkExecutorHdfsLogAppender
```
Comment (hit-lacus) [inline build/conf/spark-executor-log4j.properties:30]: `SparkExecutorHdfsLogAppender` or `SparkExecutorHdfsAppender` ?

File: build/conf/spark-executor-log4j.properties [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+log4j.appender.hdfs.identifier=${kylin.spark.identifier}
+log4j.appender.hdfs.jobName=${kylin.spark.jobName}
+log4j.appender.hdfs.project=${kylin.spark.project}
+
+log4j.appender.hdfs.rollingPeriod=5
+log4j.appender.hdfs.logQueueCapacity=5000
+#flushPeriod count as millis
+log4j.appender.hdfs.flushInterval=5000
+
+log4j.appender.hdfs.layout=org.apache.kylin.engine.spark.common.logging.SensitivePatternLayout
```
Comment (hit-lacus) [inline build/conf/spark-executor-log4j.properties:44]: Should be `org.apache.kylin.common.logging.SensitivePatternLayout` .

**Comment by hit-lacus** [comment]
image

File: core-job/src/main/java/org/apache/kylin/job/execution/AbstractExecutable.java [hunk 7]
```diff
@@ -587,7 +596,7 @@ private static boolean isRetryableException(String exceptionName) {
 
     public final String getProject() {
         if (project == null) {
-            throw new IllegalStateException("project is not set for abstract executable " + getId());
+            logger.error("project is not set for abstract executable " + getId());
```
Comment (hit-lacus) [inline core-job/src/main/java/org/apache/kylin/job/execution/AbstractExecutable.java:599]: I can not understand here.

**Comment by hit-lacus** [comment]
## Cause Analysis

image

**Comment by hit-lacus** [comment]
@zhangayqian  Please review my patch at https://github.com/zhangayqian/kylin/pull/1 , merge it if it is fine to you.

File: kylin-spark-project/kylin-spark-common/src/main/java/org/apache/kylin/engine/spark/common/logging/SparkExecutorHdfsAppender.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.UUID;
+
+public class SparkExecutorHdfsAppender extends AbstractHdfsLogAppender {
+
+    private static final long A_DAY_MILLIS = 24 * 60 * 60 * 1000L;
+    private static final long A_HOUR_MILLIS = 60 * 60 * 1000L;
+    private SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-common/src/main/java/org/apache/kylin/engine/spark/common/logging/SparkExecutorHdfsAppender.java:44]: Modify to `new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())` .

File: server-base/src/main/java/org/apache/kylin/rest/controller/JobController.java [hunk 9]
```diff
@@ -198,6 +201,22 @@ public JobInstance get(@PathVariable String jobId) {
         return new EnvelopeResponse<>(ResponseCode.CODE_SUCCESS, "", "");
     }
 
+    /**
+     * RPC Call
+     *
+     * @param sparkJobUpdateRequest
```
Comment (hit-lacus) [inline server-base/src/main/java/org/apache/kylin/rest/controller/JobController.java:207]: Totally useless comment!

**Review by hit-lacus — APPROVED** [review]
This patch looks good to me.

**Comment by hit-lacus** [comment]
What do you think of this patch, @zzcclp ?

**Comment by zzcclp** [comment]
LGTM

**Pull request merged**